staging/src/k8s.io/apimachinery/pkg/util/sets/set.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sets
+
+import (
+	"reflect"
+	"sort"
+)
+
+// SetTypes is constraints on the types of elements of the set
+type SetTypes interface {
+	~int64 | ~int | ~byte | ~string | ~int32
```

Comment:
This is not the right way to specify the permitted types. Find a way to express "can be a map key"?

Reply:
The element type of the set should be `comparable `, but the List() method requires the elements to be `ordered `, so the type constraint of the collection elements should probably be that:
```go
type SetTypes interface {
	comparable
        ordered
}
```
Cause List() result should be ordered:
```
func (s Set[T]) List() []T {
	res := make(sortableSlice[T], 0, len(s))
	for key := range s {
		res = append(res, key)
	}
	sort.Sort(res)
	return res
}
func (s sortableSlice[T]) Less(i, j int) bool { return lessSetType[T](s[i], s[j]) }
func lessSetType[T SetTypes](lhs, rhs T) bool {
	return lhs < rhs
}
```

Reply:
More details about [ordered](golang.org/x/exp/constraints).

Reply:
`Ordered` looks good. Does the go type system automatically treat `*Foo` as matching `~uintptr`? It would be handy if so.

Reply:
Should we sort the list by pointer, that sounds arbitrary

Reply:
On the other hand, one might want to have a set and doesn't care about the list function... I don't have a strong opinion on that.

Reply:
The sort is just to make the output deterministic. Usually the caller doesn't actually care about the order.

Reply:
> `Ordered` looks good. Does the go type system automatically treat `*Foo` as matching `~uintptr`? It would be handy if so.

Sadly it didn't work as we thought.

Reply:
`Comparable` seems like a suitable constraint, and I give a chance to custom sort methods for any `Comparable` type param.

Reply:
@lavalamp @apelisse PTAL.

Reply:
I just wrote a sets/generic.go for fun before asking whether it would be accepted. The solution that I came up with for this was to require only `comparable` for the set and then have a separate `List` function which adds `Ordered` as additional constraint:
```
// Generic is a set of items, implemented via map[T]struct{} for minimal memory consumption.
type Generic[T comparable] map[T]Empty
...

// List returns the contents as a sorted T slice.
//
// This is a separate function and not a method because not all types supported
// by Generic are ordered and only those can be sorted.
func List[T constraints.Ordered](g Generic[T]) []T {
	res := make(sortableSliceOfGeneric[T], 0, len(g))
	for key := range g {
		res = append(res, key)
	}
	sort.Sort(res)
	return []T(res)
}
```

Reply:
@pohly that's better than asking for a comparison function, but still is a change in function signature.

What bad thing happens if we use ordered as the main type?

Reply:
It's a bit more complicated :cry: 

A type definition does not inherit methods, so this doesn't work:
```
type String Generic[string]
```

One can add a `func (s String) List...` though.

This one is a bit better:
```
// sets.String is a set of strings.
type String struct {
	Generic[string]
}

// NewString creates a String from a list of values.
func NewString(items ...string) String {
	return String{NewGeneric(items...)}
}

// StringKeySet creates a String from a keys of a map[string](? extends interface{}).
// If the value passed in is not actually a map, this will panic.
func StringKeySet(theMap interface{}) String {
	return String{GenericKeySet[string](theMap)}
}

func (s String) List() []string {
	return List(s.Generic)
}
```

Where it fails is for code which calls `len` for the set because now something that used to be a map is a struct.

Reply:
The loose function as in @pohly's first suggestion is looking best to me so far. People already have to change their code a bit to use this.

Reply:
Actually, has it been discussed already how the generic set can be used to replace the generated sets?

I can't think of anything better than the `type String struct { Generic[string] }` approach above, but that then will break some code (`List` continues to work, `len` doesn't).

Reply:
Why not just `type String Generic[String]`?

Reply:
Such a type definition doesn't inherit any of the methods defined for `Generic[String]`. It's just the way Go is.

Reply:
> Actually, has it been discussed already how the generic set can be used to replace the generated sets?
> 
> I can't think of anything better than the `type String struct { Generic[string] }` approach above, but that then will break some code (`List` continues to work, `len` doesn't).

Many places we used as `ss := sets.String{"a", "b", "c"} `. If replace it with `type String struct { Generic[string] }`,we have to rewrite it like `ss := sets.String{ Genetic: map[string]struct{}{"1": {}, "2": {}} }`. And `len` doesn't work as @pohly said. So what about that it goes like @apelisse point.

```golang
func (s Set[T]) Insert(items ...T) Set[T] {
	for _, item := range items {
		s[item] = Empty{}
	}
	return s
}

func Insert[T comparable](set Set[T], items ...T) Set[T] {
	return set.Insert(items...)
}

// keep sets.String the same

type String Set[string]

func NewString(items ...string) String {
	return String(New[string](items...))
}

func (s String) Insert(items ...string) String {
	return String(Insert(Set[string](s), items...))
}
```

By add a bunch of generic funcs, we seem to make no breaks. And `len()` works.

Reply:
By the way, should we keep both methods？ We should allow the user to perform custom sorting for a given comparable type.

```go
func List[T constraints.Ordered](g Set[T]) []T {
	res := make(sortableSliceOfGeneric[T], 0, len(g))
	for key := range g {
		res = append(res, key)
	}
	sort.Sort(res)
	return res
}

func (s Set[T]) SortedList(less func(slice []T, i, j int) bool) []T {
	res := s.UnsortedList()
	sort.Slice(res, func(i, j int) bool {
		return less(res, i, j)
	})
	return res
}
```

Reply:
Or we can wrap the existing type:

```go
type Set[T comparable] map[T]Empty

type setWrapper[T comparable] struct {
	Set[T]
}

func wrapSet[T comparable](s Set[T]) setWrapper[T] {
	return setWrapper[T]{
		Set: s,
	}
}
```
```go
type String map[string]Empty

func (s String) Insert(items ...string) String {
	return String(wrapSet(Set[string](s)).Insert(items...))
}
```

Reply:
> Many places we used as ss := sets.String{"a", "b", "c"} 

Can you link to an example? That doesn't look like valid Go to me (missing value in the map initialization). What is used is `sets.NewString("a", "b", "c")`.

> By add a bunch of generic funcs, we seem to make no breaks. And len() works.

Yes. The downside is that we need to replicate all methods for the specific sets. I fear we must not break `len`, so I guess we'll have to use that solution. A side effect is that we can also implement the traditional list method. But I'll defer to @lavalamp  here.

> By the way, should we keep both methods？ We should allow the user to perform custom sorting for a given comparable type.

They can already to that by retrieving the unsorted slice and then invoking sort themselves. I don't think we need a specific API for that in the sets package.

Reply:
I prefer to use a wrap func to transform `String` to `Set[T]`, so we can use all methods that Set[T] had, instead of adding a bunch of funcs.

```go
type Set[T comparable] map[T]Empty

// wrap a specified set for generic Set[T].
func wrap[T comparable](s map[T]Empty) Set[T] {
	return s
}
```
```go
type String map[string]Empty

func (s String) Insert(items ...string) String {
	return String(wrap(s).Insert(items...))
}
```

Reply:
I find the name `wrap` confusing. It's not returning a wrapper, but rather the same instance cast to a different type. How about calling it `cast`?

Just a nit. Other than that such a solution should work. It's the one I mentioned above.

Reply:
Sounds good to me, I'll rename it.